modules/references/tests/server/utils.js

```diff
@@ -0,0 +1,165 @@
+    var username = defs.username + i;
+    var firstName = defs.firstName + i;
+    var lastName = defs.lastName + i;
+    var email = i + defs.email;
+    var publ = defs.hasOwnProperty('public') ? defs.public : !getRandInt(2); // public is reserved word
+    var password = defs.password || crypto.randomBytes(24).toString('base64');
+
+    return {
+      public: publ,
+      firstName: firstName,
+      lastName: lastName,
+      email: email,
+      username: username,
+      displayUsername: username,
+      password: password
+    };
+  });
+}
+
+/**
+ * Generate reference objects.
+ * @param {object[]} users - array of mongodb User
+ * @param {[number, number, object][]} referenceData - array of data for each reference
+ * @param {number} referenceData[][0] - index of userFrom in users
+ * @param {number} referenceData[][1] - index of userTo in users
+ * @param {object} referenceData[][2] - object of property: value to override default reference properties
+ */
+function generateReferences(users, referenceData) {
+  return referenceData.map(function (data) {
+    var defaultReference = {
+      userFrom: users[data[0]]._id,
+      userTo: users[data[1]]._id,
+      public: true,
+      met: !getRandInt(2),
+      hostedMe: !getRandInt(2),
+      hostedThem: !getRandInt(2),
+      recommend: ['yes', 'no', 'unknown'][getRandInt(3)]
+    };
+
+    return _.defaultsDeep({}, data[2], defaultReference);
+  });
+}
+
+/**
+ * @callback {saveDocumentsCallback}
+ * @param {error|null} error
+ * @param {object[]} documents - array of saved mongo documents
+ */
+
+/**
+ * Save documents to mongodb
+ * @param {string} collection - name of mongoose model (mongodb collection)
+ * @param {object[]} _documents - array of document data
+ * @param {saveDocumentsCallback} done
+ */
+function saveDocumentsToCollection(collection, _docs, done) {
+  var docs = _docs.map(function (_doc) {
+    var Model = mongoose.model(collection);
+    return new Model(_doc);
+  });
+
+  async.eachSeries(docs, function (doc, cb) {
+    doc.save(cb);
+  }, function (err) {
+    return done(err, docs);
+  });
+}
+
+var saveUsers = _.partial(saveDocumentsToCollection, 'User');
+var saveReferences = _.partial(saveDocumentsToCollection, 'Reference');
+
+/**
+ * Clear specified database collections
+ * @param {string[]} - array of collection names to have all documents removed
+ * @param {function} - callback
+ */
+function clearDatabase(collections, done) {
+  var models = collections.map(function (collection) { return mongoose.model(collection);});
+
+  async.eachSeries(models, function (Model, cb) {
+    Model.deleteMany().exec(cb);
+  }, done);
+}
+
+/**
+ * Sign in to app
+ * @param {object} credentials
+ * @param {string} credentials.username
+ * @param {string} credentials.password
+ * @param {object} agent - supertest's agent
+ * @param {function} done - callback
+ */
+function signIn(credentials, agent, done) {
+  agent.post('/api/auth/signin')
```

Comment:
If this returned a function it would be simpler when using it (`bind` confuses people I think).

Could also do the extraction of username/password in here to make it simpler to use, so usage would turn from:

```js
beforeEach(utils.signIn.bind(this, _.pick(_users[0], ['username', 'password']), agent)); 
```

to:

```js
beforeEach(utils.signIn(users[0], agent)); 
```

I would tend to optimize so usage (many times) is simple and make the implementation do more work (once).

Reply:
> I was about to comment usage of bind as well. If you think it's too much to refactor, don't worry about it. That said, I think it's one of those "little bit too clever" for an app maintained by volunteers with different skill levels.

So generally if I feel like using `bind`, I should rather create a function returning a function, or wrap the function into another function, right?

My (almost) only issue with functions returning functions is that they are not as general and reusable in different contexts. Another one is added complexity.

Or would you say it is easier to understand (a function returning a function) rather than (`function.bind` or `_.partial`)?

The solution covering both may be to always wrap the function in another function as needed.

There is also an interesting lodash method [`_.curry()`](https://lodash.com/docs/#curry).

Bottom line: I'm fine with following rules of thumb which suit Trustroots codebase better. Let me know the final decision on this. The refactor will be relatively easy.

Reply:
So far the username & password extraction implemented. 44af051

The rest perhaps coming soon.

Reply:
Yup, seems good as it is. I'm fine with bind :)